API recommendation based on TripGo

nighthawk opened the pull request:
Action list:

- [x] Initial clean-up and adjustments based on developer feedback
- [x] Internal review
- [x] Use http://schema.org/ImageObject for images + logos
- [x] Use off-street parking model from [FIWARE](https://fiware-datamodels.readthedocs.io/en/latest/Parking/OffStreetParking/doc/spec/index.html), similar to https://github.com/skedgo/TSP-APIs/pull/3
- [x] Use bike-hire docking station from [FIWARE](https://fiware-datamodels.readthedocs.io/en/latest/Transportation/Bike/BikeHireDockingStation/doc/spec/index.html)

Commit:
Use ISO8601 for dates not seconds since 1970

Commit:
Reviewed regions* endpoints

- use `GET`
- no `.json` in endpoint name
- no more `v` parameter
- region IDs rather than names/codes

Commit:
Clean-up

Commit:
Reuse http://schema.org/openingHours

Commit:
Clean-up

Commit:
Change routing to POST

Commit:
Move routing error to a 400 response

Commit:
Switch to OpenAPI 3.0

Commit:
Forgot to commit these

Commit:
OneOf - Hooray!

Commit:
Tweaking description, adding regions/contains

Commit:
Update lingo

operator => agency
serviceTripID => serviceID
[rR]ealtime => [rR]ealTime

Commit:
Simplify routing data model, add routing/grouped

Commit:
More clean-up

Commit:
Add tripID explicitly

Commit:
Documentation review

Commit:
Documentation review

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1,317 +1,383 @@
       parameters:
-        - name: input
-          in: body
+        - name: hashCode
```
how about using `etag` or `if-modified-since` http headers instead?

nighthawk replied:
[ETag](https://en.wikipedia.org/wiki/HTTP_ETag) sounds like a good match for this. `if-modified-since` could be harder to maintain on the backend side as the backend needs to track when what has changed when, if the response to this endpoint might depend on things like an API key.

Will try to figure out how to do this in Swagger. From how I read that Wikipedia page it means the response should have a `ETag` header while the request has an optional `If-None-Match` header, right?

nighthawk replied:
Done.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1,317 +1,383 @@
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/RegionInfoResponse'
```
I guess this is still WIP, but would be good if the format of this response is an extension of `#/components/schemas/Region`

nighthawk replied:
Good idea. Also noticed that this endpoint was returning an array, but that doesn't make sense if a single `id` is required input. Maybe we can even get rid of having both `/regions/{id}` and `/regions/{id}/details`, though that would mean that `GET /regions/{id}` would have more data for a region than doing `(GET /regions/)[id]`.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1,317 +1,383 @@
+              schema:
+                $ref: '#/components/schemas/RegionInfoResponse'
+
+  /info/agency.json:
```
I would remove `.json` in all the endpoints, and leave that to the `Accept` http header

MarianoTucat replied:
is `regionID` required? if so, how about `/regions/{id}/agencies` and removing `regionID` from the body

nighthawk replied:
Done. Also did a clean-up of the data model as both `/regions/{id}/details` and this endpoint had a concept of an "Agency" but different representation of it.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1,317 +1,383 @@
+              schema:
+                $ref: '#/components/schemas/InputError'
 
   /info/route.json:
```
same here, `/regions/{regionID}/agencies/{agencyID}/routes`?

nighthawk replied:
Hmm, good question. An issue I see with this is that regions could overlap, which makes this (and the other) endpoints harder to use. How about making agency IDs globally unique, and just use `/agencies/{agencyID}/routes`? (A stop gap measure to make them globally unique is to concatenate region ID + agency ID to make it a agency GUID, though you could get the same problem of unexpected behaviour with overlapping regions.)

nighthawk replied:
Though this would mean that IDs don't match GTFS data anymore, but I think this is fine for a transport aggregator API like this. Possibly best to make this explicit by having a `gtfsID` for everything and naming the IDs of this API ` GUID`.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -758,64 +769,68 @@ paths:
+              schema:
+                $ref: '#/components/schemas/InputError'
 
   /locationInfo.json:
```
I don't like the name and format of this endpoint. Would prefer something consistent with previous `locations` (`locations.json`). Also don't like the two possible inputs. How about splitting this one in two:

- `/locations/{identifier}/details`
- `/locations/details?lat={lat}&lng={lng}`

I'm not entirely happy with this format either

nighthawk replied:
Changing it to the first one and removing the things that passes lat+lng. It's better to use the `/locations/` endpoint for that and deal with many possible locations near that lat+lng.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -827,21 +842,26 @@ paths:
+              schema:
+                $ref: '#/components/schemas/LocationInfo'
 
   /departures.json:
```
`/regions/{id}/departures`?

nighthawk replied:
Same here about making stop IDs globally unique, then we could use `/stops/{stopID}/departures`

MarianoTucat commented on maas-api.yaml:
```diff
@@ -851,76 +871,79 @@ paths:
+                required:
+                  - embarkationStops
 
   /arrivals.json:
```
`/regions/{id}/arrivals`?

MarianoTucat commented on maas-api.yaml:
```diff
@@ -942,85 +966,92 @@ paths:
+                required:
+                  - services
 
   /service.json:
```
`/regions/{regionID}/service/{serviceID}`?

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1032,75 +1063,85 @@ paths:
+                    description: |
+                      If this service is wheelchair accessible or not. If this field is missing, this does not mean that it is not accessible, but rather that it is unknown whether it is accessible. Only if the value is present and is `false`, does it mean that the service is not wheelchair accessible.
 
   /alerts/transit.json:
```
`/regions/{regionID}/transitAlerts?`

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1110,1966 +1151,1914 @@ paths:
+          $ref: '#/components/schemas/ConfigInput'
+        from:
+          type: string
+          description: Origin coordinate as `(lat,lng)"address"` string (the `"address"` part is optional).
```
should we use [GeoCoordinates](http://schema.org/GeoCoordinates) here?

MarianoTucat replied:
I see we have `#/components/schemas/Coordinate` we can use that one

nighthawk replied:
Done, switched to using http://schema.org/GeoCoordinates

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1110,1966 +1151,1914 @@ paths:
+            List of mode identifiers that are available in this region (but not necessarily everywhere in that region).
+            Look at `modes` array in `/regions` to get more details on each mode. These are also the modes that can be
+             used as input to any routing endpoint. See
+             https:
```
should we group within a `support` object these fields:

- `streetBicyclePaths`
- `streetWheelchairAccessibility `
- `transitBicycleAccessibility `
- `transitConcessionPricing `
- `transitWheelchairAccessibility `

nighthawk replied:
Done.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1110,1966 +1151,1914 @@ paths:
+            List of mode identifiers that are available in this region (but not necessarily everywhere in that region).
+            Look at `modes` array in `/regions` to get more details on each mode. These are also the modes that can be
+             used as input to any routing endpoint. See
+             https:
```
are we going to continue using `SegmentTemplate`?

nighthawk replied:
Would be good to clean that up. I don't mind if `/routing/grouped` doesn't make it into this spec.

MarianoTucat commented on maas-api.yaml:
```diff
@@ -1110,1966 +1151,1914 @@ paths:
+            List of mode identifiers that are available in this region (but not necessarily everywhere in that region).
+            Look at `modes` array in `/regions` to get more details on each mode. These are also the modes that can be
+             used as input to any routing endpoint. See
+             https:
```
can we group `*Cost` fields (including `currencySymbol`)?

nighthawk replied:
Good idea. Added `TripMetrics` for that.

MarianoTucat reviewed (commented):
Added some comments/ideas

Commit:
Address review: Use ETag rather than hashCode parameter

Commit:
Address review: Move trip metrics to TripMetrics object

Commit:
Clean-up of region details and modes

Commit:
Tweak persisting trips

Commit:
Refactor and clean-up of retrieving agencies

Commit:
Use http://schema.org/GeoCoordinates

Commit:
More lat/lng => latitude/longitude conversion

fersagui commented on maas-api.yaml:
```diff
@@ -1110,1966 +1120,1834 @@ paths:
+          - bikePod
+          - modeInfo
+
+    CarParkInfo:
```
you could use the updated car park schema from https://github.com/skedgo/TSP-APIs/pull/3/files

Commit:
Update /locations

- no ".json" in URLs
- remove cell ID variant
- unify output

Commit:
Fix bad format

Commit:
Review /geocode

Commit:
Reviewed URLs and grouping

nighthawk commented:
@MarianoTucat , I think I addressed all your review comments. Do you want to have another look over it?

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/Region'
+
+  /regions/{id}/details:
```
can we merge this into `/regions/{id}` and always return all details?

nighthawk replied:
From an implementation side, rather not:

1. `details` contains information that's possibly dynamic. Just looking at our backend, having all that available for all regions for a single API call is not trivial. 
2. `/regions` output in our case is already very large, also adding all those details for every regions makes the output massive.

The idea here is to have `/regions` return high level coverage details ("what regions do you support, with what *modes* in a region") while `details` include much more granular information of what *data* is available in each region.

jayenashar replied:
i understand `/regions` returns a summary for all/multiple regions.  i just don't get why `/regions/{id}` would return a summary.

nighthawk replied:
Done. One endpoint less. Hooray.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/InputError'
+
+  /routing/grouped:
```
can we merge this with `/routing` and add `group` as a parameter?

nighthawk replied:
No, different outputs schemas. It's discouraged for the same API endpoint for the same response type and the same HTTP status code to return different output schemas depending on the input.

jayenashar replied:
got it.  can we not use `/` as a separator then?  maybe `/trips` and `/trip-groups`?

jayenashar replied:
`/services/realtime` isn't working for me either.  maybe `/realtime/services` and `/realtime/alerts`?

nighthawk replied:
Sounds good to me

nighthawk replied:
Done

nighthawk replied:
Missed the other comment: Why not use `/` as a separator?

jayenashar replied:
Just doesn't feel like a hierarchy.  /trips (/routing) and /trip-groups (/routing/grouped) seem like they should be at the same level.

jayenashar commented on maas-api.yaml:
```diff
@@ -564,28 +686,19 @@ paths:
+              schema:
+                $ref: '#/components/schemas/Trip'
 
   /trip/{id}/save:
```
shall we remove `/save`, `/update`, and `/planned` from the API specs as they are in the response of `/trip/{id}`?  can we unify them to all have the same response schema?  for `/update`, instead of the `If-None-Match` header, we may need to include a query parameter in the `updateURL` response field?

nighthawk replied:
I'm not sure, I follow. They serve different use cases, so I do not see the advantage of unifying them.

nighthawk replied:
I added the `If-None-Match` as a replacement to passing a query parameter to use a more HTTP standardway of achieving the same thing.

jayenashar replied:
they do serve different use cases, but they all are actions on the same trip.  what i was saying before is that they are already specified in the response of other URLs so i wasn't sure if they need to be documented separately.  They do need to be documented separately if they have different response schemas, so I wanted to unify them.  Also they need to be documented if they require a special header or post body.

this is what i'm thinking now:

1. could we remove the saveURL, updateURL, plannedURL from the other responses and rely on developers to use the defined url format in the API?
2. can we change trip saving to be a PUT to `/persistentURLs` with a PUT body of `tripID`?  (sorry i don't know the URL scheme for persistent URLs)
3. can we merge trip updating with getting?  `If-None-Match` would then be documented under `/trip/{id}`
4. can we make trip planned to be a POST to `/trip/{id}`?  then we can have more actions in the future without creating a new URL for each action.

btw, if i save a trip, does the `updateURL` returned work forever?  (it's not clear from the docs)

nighthawk replied:
Ok, makes more sense now.

1. I don't see any issue with documenting these endpoints, even if trips already have the URL for them in their reply. It's good practice to document endpoints even if you don't construct the URLs themselves yourselves (which is the only info contained in the trips). There's a recommendation for APIs to provide follow-up URLs in the responses rather than developers creating URLs themselves.
2. Interesting idea though it means we can't just have the URL in the trip response anymore and let users GET on that. I guess you don't want both `GET trips/{id}/save` and `PUT /persistentURLs`?
3. Good idea, though we'd lose the information of whether the trip has any real-time data or not (via the presence of `updateURL`), but I don't mind that this then implicitly means that every trip could have real-time data.
4. I prefer not as I would read `POST /trip/{id}` as creating/storing a trip, not adding something to an existing trip, so I prefer the current format.

In general, I prefer many simple endpoints over fewer but more complex or overloaded endpoints. But that's also a discussion to be had with the MaaS Alliance working group.

nighthawk replied:
I've addressed the third point by merging `trip/{id}/update` into `trip/{id}`

jayenashar replied:
1. You are right, we should document all endpoints, even ones that are part of a hypermedia API.
2. You understood me correctly.  Just trying to get closer to CRUD, especially in the case where we are saving some data.
3. Thanks
4. How about `PATCH`?

Is there someone in MaaS Alliance we can/should tag in this pull request, so they have input into the proposed API?  Or is that too many chefs in the kitchen?

nighthawk replied:
I've merged in this PR here and submitted it as a new PR to the repo of the MaaS Alliance: https://github.com/maas-alliance/apis/pull/6. So all follow-up discussions or suggestions is best to go there, or if it's a specific proposal as a new PR in this repo.

On 2: I agree that `GET /trips/{id}/save` is not ideal and doesn't match CRUD well. On 4: I prefer to stick with `GET, POST, PUT, DELETE` as it gets a bit fuzzy when you use the HTTP verbs.

Anyway, there's also a call with the working group on Thursday. Will keep you posted if there's any news.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/Trip'
 
   /trip/{id}:
```
other endpoints like `/locations/{id}` and `/regions/{id}` are plural.  should we use `/trips/{id}` here?

nighthawk replied:
Good catch, will change that.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/Agency' 
+
+  /agencies/{aid}/routes:
```
can we move this to `/routes` and have a parameter for `aid`?

nighthawk replied:
I prefer the current way (also suggested by Mariano), as the hierarchical structure of the URL follows REST.

nighthawk replied:
API convention is that `GET /routes/{some id}` should give the route of that ID. Or do you just mean having `GET /routes/?agencyId={aid}`?

nighthawk replied:
Played around with it and agree that this looks nicer overall in the specs. Will change it. @MarianoTucat, is that okay with you?

MarianoTucat replied:
yes, sounds good to me

jayenashar replied:
oh i don't like the GET & POST for the same URL taking different parameters.  can we do something about that?  (like allow POST only for v1)

same question for `/services`

nighthawk replied:
It's pretty common to have GET & POST to the same URL, taking different parameter, but ideally returning the same response. That's what we do here.

Will keep it like this and see what feedback we get on this from the MaaS Alliance working group.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/RoutesResponse'
+
+  /agencies/{aid}/routes/{rid}/services:
```
can we move this to `/services` and have parameters for `aid` and `rid`?

nighthawk replied:
Done

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+              schema:
+                $ref: '#/components/schemas/RoutesResponse'
+
+  /agencies/{aid}/services/{sid}:
```
can we move this to `/services` and have parameters for `aid` and `sid`?

nighthawk replied:
To keep it most consistent with other endpoints, this should rather be `GET /services/{sid}?agencyId={aid}`

nighthawk replied:
Rationale: Otherwise this would clash with `GET /services` when you provide agency ID + routes ID, which returns a different schema. Can't overload `GET` calls to a single endpoint like that, which would be a bit weird anyway.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+                    description: |
+                      If this service is wheelchair accessible or not. If this field is missing, this does not mean that it is not accessible, but rather that it is unknown whether it is accessible. Only if the value is present and is `false`, does it mean that the service is not wheelchair accessible.
+
+  /routes/filter:
```
can we just make this `/routes`?

nighthawk replied:
Yep, with your other suggestion that also means that `GET /routes` and `POST /routes` have the same response schema. Hooray.

jayenashar commented on maas-api.yaml:
```diff
@@ -1,555 +1,677 @@
+                $ref: '#/components/schemas/InputError'
 
-  /info/service.json:
+  /services/filter:
```
can we just make this `/services`?

nighthawk replied:
Yep, good idea. Also same response schema here for `GET` and `POST` that way.

jayenashar reviewed (commented):
mostly suggesting changing endpoints

Commit:
Address Jayen's comments

Commit:
Also adopt /alerts to new style

Commit:
Simpler trip segments

Commit:
Rename real-time servlets

/services/realtime => /realtime/services
/alerts => /realtime/alerts

Commit:
Make /regions/{id} give full details

Commit:
Merge trip/{id}/update into trip/{id}

Commit:
Use http://schema.org/ImageObject

Commit:
Aligning with FIWARE + schema.or

- BikePodInfo => Fiware's BikeHireDockingStation
- CarParkInfo => Fiware's OffStreetParking
- CompanyInfo => https://schema.org/provider

Commit:
More external docs
